bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/DataClientVeneerApi.java

```diff
@@ -229,8 +250,96 @@ protected void onCompleteImpl() {
+    private static final int MIN_BYTE_BUFFER_SIZE = 100 * 1024 * 1024;
+    private static final double DEFAULT_BYTE_LIMIT_PERCENTAGE = .1;
+    private static final RowResultAdapter RESULT_ADAPTER = new RowResultAdapter();
 
+    private final Meter scannerResultMeter =
+        BigtableClientMetrics.meter(BigtableClientMetrics.MetricLevel.Info, "scanner.results");
+    private final Timer scannerResultTimer =
+        BigtableClientMetrics.timer(
+            BigtableClientMetrics.MetricLevel.Debug, "scanner.results.latency");
+
+    private ServerStream<Result> serverStream;
+    private ByteString lastSeenRowKey = ByteString.EMPTY;
+    private Boolean hasMore = true;
+    private final Queue<Result> buffer;
+    private final Query.QueryPaginator paginator;
+    private final paginatorFunction wrapper;
+    private final int refillSegmentWaterMark;
+
+    private static final long maxSegmentByteSize =
+        (long)
+            Math.max(
+                MIN_BYTE_BUFFER_SIZE,
+                (Runtime.getRuntime().totalMemory() * DEFAULT_BYTE_LIMIT_PERCENTAGE));
+    private long currentByteSize = 0;
+
+    PaginatedRowResultScanner(
+        Query.QueryPaginator paginator, int pageSize, paginatorFunction wrapper) {
+      this.paginator = paginator;
+      this.wrapper = wrapper;
+      this.buffer = new ArrayDeque<>();
+      this.refillSegmentWaterMark = (int) Math.max(1, pageSize * WATERMARK_PERCENTAGE);
+      this.serverStream = this.wrapper.func(this.paginator);
+    }
+
+    @Override
+    public Result next() {
+      try (Context ignored = scannerResultTimer.time()) {
+        if (this.buffer.size() < this.refillSegmentWaterMark
+            && this.serverStream == null
+            && hasMore) {
+          this.serverStream = this.wrapper.func(this.paginator);
+        }
+        if (this.buffer.isEmpty() && this.serverStream != null) {
+          this.waitReadRowsFuture();
+        }
+        scannerResultMeter.mark();
+        Result result = this.buffer.poll();
+        if (result != null) {
+          currentByteSize -= Result.getTotalSizeOfCells(result);
+        }
+        return result;
+      }
+    }
+
+    @Override
+    public void close() {
+      if (this.serverStream != null) {
+        this.serverStream.cancel();
+        this.serverStream = null;
+      }
+    }
+
+    public boolean renewLease() {
+      throw new UnsupportedOperationException("renewLease");
+    }
+
+    private void waitReadRowsFuture() {
+      Iterator<Result> iterator = this.serverStream.iterator();
+      while (iterator.hasNext()) {
+        Result result = iterator.next();
+        this.buffer.add(result);
+        if (result == null || result.rawCells() == null) {
+          continue;
+        }
+        this.lastSeenRowKey = RESULT_ADAPTER.getKey(result);
+        this.currentByteSize += Result.getTotalSizeOfCells(result);
+        if (this.currentByteSize >= maxSegmentByteSize) {
```

Comment:
Can we also add a test to test the behavior when maxSegmentByteSize is reached? We can make maxSegmentByteSize accessible to test and set it manually to a lower value.

Reply:
And I think here we need to cancel the serverStream

Reply:
Done